topdown/http.go

```diff
@@ -568,3 +615,331 @@ func parseTimeout(timeoutVal ast.Value) (time.Duration, error) {
 		return timeout, builtins.NewOperandErr(1, "'timeout' must be one of {string, number} but got %s", ast.TypeName(t))
 	}
 }
+
+func getBoolValFromReqObj(req ast.Object, key string) (bool, error) {
+	result := false
+	var err error
+
+	value := req.Get(ast.StringTerm(key))
+	if value != nil {
+		result, err = strconv.ParseBool(value.String())
+		if err != nil {
+			return false, err
+		}
+	}
+	return result, nil
+}
+
+type persistentCacheValue struct {
+	value ast.Value
```

Comment:
Since we're only storing the size in bytes and ast.Value on the cache elements we end up having to implement all kinds of conversion logic to pull the caching headers out of the ast.Value response just to perform some simple comparisons for invalidation.

I think this code would be much simpler if we kept the information necessary to validate the cache expiry on the struct itself.

Reply:
Refactored code such that `persistentCacheValue` holds information to check response freshness and re-validate if needed.